src/storage-client/src/controller.rs

```diff
@@ -1220,6 +1240,30 @@ where
         Ok(self.state.persist_write_handles.append(commands))
     }
 
+    async fn finalize(&mut self, identifiers: Vec<GlobalId>) -> Result<(), StorageError> {
+        loop {
+            let updates = identifiers
+                .iter()
+                .map(|id| (id.clone(), Vec::new(), Antichain::new()))
+                .collect();
+
+            let res = self
+                .append(updates)
+                .expect("[] not beyond upper")
+                .await
+                .expect("one-shot dropped while waiting synchronously");
+
+            if let Err(StorageError::InvalidUppers(_)) = res {
```

Comment:
Maybe add a comment saying roughly: "This can happen because of concurrent writes, we want to make sure that our finalization _does_ go through, so we retry indefinitely."

Also, though: I'm afraid the "advance to []" might work for some identifiers, but not for all and we might then be trapped in a retry loop?

Reply:
> Also, though: I'm afraid the "advance to []" might work for some identifiers, but not for all and we might then be trapped in a retry loop?

Oh interesting! When would that be the case?

Reply:
I don't know, but aren't all the "advance to []" calls `compare_and_append` calls in the end? So they might fail or complete independently. Then when we retry the next time around the loop, this will keep failing for those shards where we already successfully advanced to `[]`?

Reply:
Further discussion in https://materializeinc.slack.com/archives/C01CFKM1QRF/p1671636579620219

I'm not quite sure what the right change here is, but I think we'll want to make some changes to the persist worker and/or compare_and_append for tombstoned behavior. offhand feels like advancing the upper to `[]` when it's already `[]` should succeed